### PR TITLE
rework LDAP authorization plugins to cease being DN centric

### PR DESCRIPTION
**Pull request by vladak**
This set of changes is needed for more generic user handling. Currently all the LDAP plugins depend on the fact that the header decoder sets LDAP DN into `User.username`. With some deployments of SAML for instance, this is no longer the truth as it passes just the e-mail address of the user.

The most visible configuration change is for `LdapUserPlugin`. It no longer takes `objectclass` parameter (to construct LDAP filter) but now takes the filter explicitly and expands `User` properties (similarly to `LdapFilterPlugin`). So far it only expands `username` and `guid` properties of `User`. Also, one can specify `useDN` boolean value to indicate whether the `User.username` emitted by the decoder in use should be treated as LDAP DN. If false (the default), it will extract the DN from the search results and store it as `LdapUser` property. This property can be used by LDAP plugins down the stack to refer to the user directly - this is obviously the most useful for `LdapAttrPlugin` which just matches  the LDAP attributes of given object.

Here's a sample of `LdapUserPlugin` config:
```xml
            <void method="add">
                <object class="org.opengrok.indexer.authorization.AuthorizationPlugin">
                    <void property="name">
                        <string>opengrok.auth.plugin.LdapUserPlugin</string>
                </void>
                <void property="flag">
                    <string>REQUISITE</string>
                </void>
                <void property="setup">
<!--
                    <void method="put">
                        <string>useDN</string>
                        <boolean>true</boolean>
                    </void>
-->
                    <void method="put">
                        <string>filter</string>
                        <string>(&amp;(objectclass=person)(mail=%username%))</string>
                    </void>
                    <void method="put">
                         <string>attributes</string>
                         <string>mail,uid</string>
                    </void>
                </void>
                </object>
            </void>
```

The configuration of the other LDAP plugins did not change.

I used the opportunity to clean up some of the neighbouring code.

### CODE REVIEW

**Comment by vladak** [comment]
The trouble with the mail-to-DN approach is that the `LdapUserPlugin` e-mail based search can result in multiple objects like so:
```
$ ldapsearch -h someldap.foo.com -b dc=foo,dc=com '(&(objectclass=posixAccount)(mail=vladimir.kotal@foo.com))' uid 
version: 1                                                                      
dn: cn=VLADIMIR_KOTAL,l=amer,dc=foo,dc=com                                      
uid: vkotal                                                                     
                                                                                
dn: cn=grok,l=amer,dc=foo,dc=com                                                
uid: grok                                                                       
                                                                                
dn: cn=wsmirror,l=amer,dc=foo,dc=com                                            
uid: wsmirror
```

but for now I find this acceptable.

File: plugins/src/opengrok/auth/plugin/LdapUserPlugin.java [hunk 1]
```diff
@@ -103,41 +101,59 @@ protected boolean sessionExists(HttpServletRequest req) {
...
+     */
+    protected String expandFilter(User user) {
+        String filter = ldapFilter;
+
+        filter = filter.replaceAll("(?<!\\\\)%username(?<!\\\\)%", user.getUsername());
+        filter = filter.replaceAll("(?<!\\\\)%guid(?<!\\\\)%", user.getId());
+
+        filter = filter.replaceAll("\\\\%", "%");
+
+        return filter;
```
Comment (tulinkry) [inline plugins/src/opengrok/auth/plugin/LdapUserPlugin.java]: duplicated code

File: plugins/src/opengrok/auth/entity/LdapUser.java [hunk 2]
```diff
@@ -34,17 +34,20 @@
  */
 public class LdapUser implements Serializable {
 
+    private String DN;
```
[the remark this answers is not in the document]
Comment (vladak) [reply plugins/src/opengrok/auth/entity/LdapUser.java]: It's an acronym for Distinguished Name. Given it is used for storing username as well, the naming should be different.

File: plugins/src/opengrok/auth/plugin/util/FilterUtil.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+ * file and include the License file at LICENSE.txt.
+ * If applicable, add the following below this CDDL HEADER, with the
+ * fields enclosed by brackets "[]" replaced with your own identifying
+ * information: Portions Copyright [yyyy] [name of copyright owner]
+ *
+ * CDDL HEADER END
+ */
+
+/*
+ * Copyright (c) 2016, 2019 Oracle and/or its affiliates. All rights reserved.
```
[the remark this answers is not in the document]
Comment (vladak) [reply plugins/src/opengrok/auth/plugin/util/FilterUtil.java]: I copied the code from elsewhere, so retaining the original year.

File: plugins/src/opengrok/auth/plugin/LdapAttrPlugin.java [hunk 4]
```diff
@@ -110,16 +110,22 @@ public void fillSession(HttpServletRequest req, User user) {
...
+                String dn = ldapUser.getDN();
+                if (dn != null) {
+                    LOGGER.log(Level.FINEST, "searching with dn={0}", dn);
+                    AbstractLdapProvider.LdapSearchResult<Map<String, Set<String>>> res;
+                    if ((res = getLdapProvider().lookupLdapContent(dn, new String[]{ldapAttr})) == null) {
+                        LOGGER.log(Level.WARNING, "cannot lookup attributes {0} for user {1} (LDAP provider: {2})",
+                                new Object[]{ldapAttr, user, getLdapProvider()});
+                    }
+
+                    records = res.getAttrs();
```
[the remark this answers is not in the document]
Comment (vladak) [reply plugins/src/opengrok/auth/plugin/LdapAttrPlugin.java]: good catch, adding return.

File: plugins/src/opengrok/auth/plugin/LdapUserPlugin.java [hunk 5]
```diff
@@ -103,41 +103,58 @@ protected boolean sessionExists(HttpServletRequest req) {
...
+        String dn;
+        if (ldapFilter != null) {
+            expandedFilter = expandFilter(user);
+        }
         try {
-            if ((records = getLdapProvider().lookupLdapContent(null, filter, attributes)) == null) {
-                LOGGER.log(Level.WARNING, "failed to get LDAP attributes ''{3}'' for user ''{0}'' " +
+            AbstractLdapProvider.LdapSearchResult<Map<String, Set<String>>> res;
+            if ((res = getLdapProvider().lookupLdapContent(useDN ? user.getUsername() : null,
+                    expandedFilter, attributes.toArray(new String[attributes.size()]))) == null) {
```
[the remark this answers is not in the document]
Comment (vladak) [reply plugins/src/opengrok/auth/plugin/LdapUserPlugin.java]: yep, will be replaced by default filter along the way.

**Comment by vladak** [comment]
In order for this to be useful, I needed to fix another limitation. The stack definition of internal deployment used `LdapUserPlugin` multiple times, once on the top level, once in a sub-stack. The problem is that the substack invocation of the plugin rewrites the session attribute (which holds `LdapUser` object) that can be then reused later on in different substack. This can result in object not found when `LdapAttrPlugin` is used because it relies on the `LdapUser` attributes when searching. Thus, I allowed to have multiple `LdapUser` session attributes by specifying instance number.

**Comment by vladak** [comment]
Once this is in, my plan is to create new wiki page with detailed description of the plugins and their configuration.

**Comment by vladak** [comment]
New wiki: https://github.com/oracle/opengrok/wiki/Authorization-plugins

**Pull request merged**